### ISSUE TITLE
Building on macOS

### ISSUE THREAD

**Issue by akx** [bintracker/bintracker#20]
On my macbook, I can get a `bintracker` binary with

```
$ gh repo clone bintracker/bintracker
$ cd bintracker
$ brew install chicken sqlite3 tcl-tk  # (maybe other things)
$ env CHICKEN_C_LIBRARY_PATH=/opt/homebrew/opt/sqlite/lib CHICKEN_C_INCLUDE_PATH=/opt/homebrew/opt/sqlite/include chicken-install -v args base64 bitstring comparse coops list-utils matchable pstk s11n simple-md5 sqlite3 stb-image stb-image-write srfi-1 srfi-4 srfi-13 srfi-14 srfi-18 srfi-69 stack test typed-records web-colors
$ cd build
$ make -j4
```
Running `./bintracker` then results in
```
Error: (open-input-file) cannot open file - No such file or directory: "config/keymaps/en.keymap"

	Call history:

	../bt-maths.scm:9: chicken.load#load-extension
	../bintracker-core.scm:9: chicken.load#load-extension
	../bt-state.scm:13: chicken.load#load-extension [repeated]
	../bt-types.scm:8: chicken.load#load-extension [repeated]
	../bt-types.scm:127: scheme#call-with-input-file

```
which is understandable (and promising).

cd'ing back to the project root and trying `./build/bintracker` results in
```
Error: (import) during expansion of (import ...) - cannot import from undefined module: mdal

	Call history:

	<syntax>	  (##core#require library chicken.base)
	<syntax>	  (chicken.load#load-unit (##core#quote library) (##core#quote #f) (##core#quote #f))
	<syntax>	  (##core#quote library)
	<syntax>	  (##core#quote #f)
	<syntax>	  (##core#quote #f)
	<syntax>	  (##core#begin (##core#require expand chicken.syntax))
	<syntax>	  (##core#require expand chicken.syntax)
	<syntax>	  (chicken.load#load-unit (##core#quote expand) (##core#quote #f) (##core#quote #f))
	<syntax>	  (##core#quote expand)
	<syntax>	  (##core#quote #f)
	<syntax>	  (##core#quote #f)
	<eval>	  (chicken.load#load-unit (##core#quote library) (##core#quote #f) (##core#quote #f))
	<eval>	  (chicken.load#load-unit (##core#quote library) (##core#quote #f) (##core#quote #f))
	<eval>	  (chicken.load#load-unit (##core#quote expand) (##core#quote #f) (##core#quote #f))
	<syntax>	  (import-syntax scheme chicken.base chicken.bitwise chicken.blob chicken.condition chicken.fixnum chi...
	<syntax>	  (##core#undefined)	<--
```
which sounds like `mdal.so` (which is in `build`) can't be loaded.

I'll keep investigating...

**Comment by akx** [comment]
One ugly `cp build/*.so .` later:

<img width="608" alt="Screenshot 2023-09-06 at 14 22 34" src="https://github.com/bintracker/bintracker/assets/58669/41849d7c-d942-43f9-ac30-3402e30ea8cb">

Select Engine doesn't look good though:

<img width="946" alt="Screenshot 2023-09-06 at 14 23 16" src="https://github.com/bintracker/bintracker/assets/58669/b79bb246-c479-4f82-9562-3cf90581fd13">

**Comment by akx** [comment]
Ah... Instead of copying the .so's into the project directory, one needs to `cp -r ../config/ .` so there's a `build/config` directory, and then run `bintracker` from the `build` directory.

Then engines show up... but trying to use e.g. `tiatune` (since `atari2600` shouldn't need a ROM) yields

```
Attempt 1: Failed to connect to emulator
Attempt 2: Failed to connect to emulator
```
(with a blank `mame` window in the background) and a hanging Select Engine window.

**Comment by ctsrc** [comment]
@akx regarding mame, did you install the cask version or the non-cask version?

I noticed that the cask version is from a separate project something like mame osx, and the version number of that one seems much older than the non-cask version.

So if you installed the cask version, try the non-cask version instead.

**Comment by akx** [comment]
Non-cask, formula.

**Comment by akx** [comment]
That said, hey... should `lua` be a dependency there or something? (Is it possible to build Mame without lua bootscript support?)

**Comment by utz82** [comment]
Hi @akx, many thanks for trying to hack a path through the jungle.

Unfortunately I'm completely unfamiliar with macOS, so I'm afraid I can't be of too much help here.

Generally the idea is to run `bintracker` from the `build` directory, as you said. Normally, the `make` receipe should copy the `config` dir, as well as a number of other dirs. Looks like this isn't actually happening. Any ideas what's going wrong here?

The emulator not starting up could be down to two things. Either Bintracker<->Mame communication via TCP isn't working as intended, or Mame can't find some files. Is the `mame-bridge` directory present in `build`? The strange thing is  that it hangs after 2 connection attempts. Normally up to 10 tries should be performed.

Mame can indeed be build without Lua support, though Lua support is enabled by default, and no system-wide Lua installation is required. Probably a good idea to mention it in the build instructions though.

**Comment by akx** [comment]
Yeah, so as far as I can see, mame does start (after 2 attempts), but Something(tm) happens and the connection between bintracker and mame fails.

Should mame's stdout/stderr prints (are there any in general?) appear in the bintracker console? Can I pepper print statements of some sort into the bridge lua file?

**Comment by akx** [comment]
Oh, it... apparently doesn't hang if I open a demo tune! Trying to create a new one reliably fails, but I could actually get sound out of the tiatune demo 🎉 

<img width="1183" alt="Screenshot 2023-09-08 at 22 02 29" src="https://github.com/bintracker/bintracker/assets/58669/f5570e0c-63e3-4bee-9874-37a2033943ca">

**Comment by utz82** [comment]
Thanks for the reports. That's kind of strange, but promising at least. When the module shows up on creating a new track, does it show the edit cursor?

There seem to be some platform differences as to where Mame's stdout/stderr ends up. If regular print doesn't work, try the local rprint instead, which should always print to the bintracker console. The latter only prints plain strings, though.

However, I suspect the problem isn't actually with mame-bridge, but instead something is going wrong between the bintracker backend and the gui. Unfortunately, as I mentioned in #18 I'm currently too busy to look into this, my apologies for that. If you're up for digging further, by all means please do. You could try to play around with https://github.com/bintracker/bintracker/blob/97a0ec784398fd95ad52d2e6c6ed42e3ac32319f/bt-gui-lolevel.scm#L1533. Perhaps comment out L1535 + L1607, or replace the `tk` in L1545 with `(ui-parent d)`.
